### ISSUE TITLE
`@map` on the query level

### ISSUE THREAD

**Issue by AlaaZorkane** [prisma/prisma#6725]
## Problem

Hey! Would love it if the query builder API introduces a way to `@map` on the query level. Per example change an `id` field on the response object to be `databaseId` but only on a single query. It's possible via a raw query using `FORMAT` but it would be nicer to do so on the query level. I think this would hugely increase performance on some cases when you have to loop on a `findMany` response to change the name of certain fields.

## Suggested solution

Take a select query:
```js
const users = prisma.user.findMany({ take: 10 })
```
which is the equivalent to
```sql
SELECT `main`.`User`.`id`, `main`.`User`.`createdAt`, `main`.`User`.`updatedAt`,
`main`.`User`.`email`, `main`.`User`.`name` FROM `main`.`User` LIMIT 10
```

To map only for this specific query you'd have to

```js
const newUsers = users.map(x => ({...x, databaseId: x.id, id: undefined }))
```
which can be done with sql via the `as` operator like so (which is **much much** faster and scales very well):
```sql
SELECT `main`.`User`.`id` as databaseId, `main`.`User`.`createdAt`, `main`.`User`.`updatedAt`,
`main`.`User`.`email`, `main`.`User`.`name` FROM `main`.`User` LIMIT 10
```

Suggested API:
```js
const users = prisma.user.findMany({ take: 10, map: { id: "databaseId" } })
```

## Alternatives

Maybe not the keyword `map` but something else to not confuse people?
Also with the suggested API I didn't really give it much thought on how would it be property typed but I am pretty sure it's possible, either way that would be I guess one of the challenges.

**Comment by smakosh** [comment]
wow this would be so useful, thanks alaa for this great feature request.

**Comment by janpio** [comment]
Did I understand correctly that you want to have `id` in the database and the Prisma schema data model, but the returned Javascript object should contain the data as `databaseId`?
Can you give insight why using `@map` on the schema level (to make it `databaseId` there and the Client results, but `id` in the database) is not an option?

**Comment by AlaaZorkane** [comment]
`@map` will make all the queries return `databaseId` instead of `id`, and will also change the property name for other methods (input and output), on the other hand, what I am proposing works basically for a specific **single** query and only for the returned data (not the input data).

Motivation would be to give people more control on the shape of the returned data, in my case I want this so I can make the `id` a `databaseId` on certain parts of my app

**Comment by janpio** [comment]
Thanks, understood and a valid request to be able to do that on a query level. Our biggest problem would probably be our type system here, as that of course uses the `id` from the schema to type these objects. But that is of course secondary to your request 👍

**Comment by AlaaZorkane** [comment]
Yes, as I mentioned in the initial issue, programmatically typing the result object depending on the user's input will be one of the challenges, I am thinking some sort of utility function like with `Select` but yeah I will explore on that if I came up with anything I will make sure to share it here, either way, I didn't delve too much into the prisma query constructor code but I think it would be slightly easier on that part. Thank you!

**Comment by AlaaZorkane** [comment]
Okay so here is what I was able to come up with (with a help from a friend of mine and a kind stranger in typescript's discord)
```ts
// This is a wrapper around the methods that will be using map,
// required to be able to read the args (and the value of `map`) in the form of `as const`
declare function find<A>(x: Narrow<A>): A;

type Cast<A, B> = A extends B ? A : B;

type Narrowable =
    | string
    | number
    | bigint
    | boolean;

type Narrow<A> = Cast<A,
    | []
    | (A extends Narrowable ? A : never)
    | ({ [K in keyof A]: Narrow<A[K]> })
>;

// Our prisma model - works with complex types say per example posts etc
type User = {
    id: string;
    name: string;
    isVerified: boolean;
}

const args = find({ map: { id: "databaseId" } });

type y = typeof args["map"];

type Mapping<T> = { [K in keyof T]?: string };

type Replace<T, U extends Mapping<T>> = {
  [K in keyof T as K extends keyof U ? string & U[K] : K]: T[K]
};

type ControlledUserOutput = Replace<User, y>

const dbOutput: ControlledUserOutput = {
    databaseId: "something",
    name: "primsic",
    isVerified: true,
}
```

**Comment by ratiomark** [comment]
So, currently, when fetching related data (e.g., card with box details), the data is nested, requiring additional mapping to flatten the structure:
```typescript
async getCardsWithBoxIndexAndTitle(userId: User['id']) {
  const cards = await this.prisma.card.findMany({
    where: { userId },
    include: {
      box: {
        select: {
          index: true,
          title: true,
        }
      }
    }
  })

  // Further mapping to flatten the structure
  const cardsWithBoxIndexAndId = cards.map((card) => ({
    ...card,
    boxIndex: card.box.index,
    boxTitle: card.box.title,
  }))

  return cardsWithBoxIndexAndId;
}

```

### Suggested Feature:

1. **Mapping** (map option): This would allow direct renaming and restructuring of fields from related entities in the query itself. For instance, map: { index: 'boxIndex', title: 'boxTitle' } would map box's index and title directly to boxIndex and boxTitle in the card entity.

2. **Node Collapsing** (removeNode option): This feature would enable the removal of the intermediate node (like box) and integrate its fields directly into the parent entity. This way, we could access card.boxIndex directly instead of card.box.boxIndex.

```ts
async getCardsWithBoxIndexAndId(userId: User['id']) {
  const cardsWithBoxIndexAndId = await this.prisma.card.findMany({
    where: { userId },
    include: {
      box: {
        select: {
          index: true,
          title: true,
        },
        map: {
          index: boxIndex,
          title: boxTitle,
        },
        removeNode: true // or collapseNode something like that
      }
    },
  });

  return cardsWithBoxIndexAndId;
}
 ```

These features would allow for more streamlined and efficient data handling, especially in cases where a flattened data structure is preferable. 

@janpio, I've also shared my thoughts in the thread [Change response field names](https://github.com/prisma/prisma/issues/4433). I believe a separate discussion may not be necessary for this suggestion, but it's important for the Prisma team to consider the idea of streamlining the structure of database query responses.

Prisma ❤
